pkg/build/api/validation/validation_test.go

```diff
@@ -40,6 +40,161 @@ func TestBuildValidationSuccess(t *testing.T) {
+			},
+			Status: buildapi.BuildStatus{
+				Phase: buildapi.BuildPhaseNew,
+			},
+		},
+	}
+	for _, build := range builds {
+		if result := ValidateBuild(&build); len(result) > 0 {
+			t.Errorf("Unexpected validation error returned %v", result)
+		}
+	}
+
+	badBuild := &buildapi.Build{
+		ObjectMeta: kapi.ObjectMeta{Name: "buildid", Namespace: "default"},
+		Spec: buildapi.BuildSpec{
+			Source: buildapi.BuildSource{},
+			Strategy: buildapi.BuildStrategy{
+				DockerStrategy: &buildapi.DockerBuildStrategy{},
+			},
+			Output: buildapi.BuildOutput{
+				To: &kapi.ObjectReference{
+					Kind: "DockerImage",
+					Name: "repository/data",
+				},
+			},
+		},
+		Status: buildapi.BuildStatus{
+			Phase: buildapi.BuildPhaseNew,
+		},
+	}
+	if result := ValidateBuild(badBuild); len(result) == 0 {
+		t.Error("An error should have occurred with a DockerStrategy / no source combo")
+	}
+
+}
+
+func TestBuildConfigEmptySource(t *testing.T) {
+	buildConfigs := []buildapi.BuildConfig{
+		{
+			ObjectMeta: kapi.ObjectMeta{Name: "config-id", Namespace: "namespace"},
+			Spec: buildapi.BuildConfigSpec{
+				BuildSpec: buildapi.BuildSpec{
+					Source: buildapi.BuildSource{},
+					Strategy: buildapi.BuildStrategy{
+						SourceStrategy: &buildapi.SourceBuildStrategy{
+							From: kapi.ObjectReference{
+								Kind: "DockerImage",
+								Name: "myimage:tag",
+							},
+						},
+					},
+					Output: buildapi.BuildOutput{
+						To: &kapi.ObjectReference{
+							Kind: "DockerImage",
+							Name: "repository/data",
+						},
+					},
+				},
+			},
+		},
+		{
+			ObjectMeta: kapi.ObjectMeta{Name: "config-id", Namespace: "namespace"},
+			Spec: buildapi.BuildConfigSpec{
+				BuildSpec: buildapi.BuildSpec{
+					Source: buildapi.BuildSource{},
+					Strategy: buildapi.BuildStrategy{
+						CustomStrategy: &buildapi.CustomBuildStrategy{
+							From: kapi.ObjectReference{
+								Kind: "DockerImage",
+								Name: "myimage:tag",
+							},
+						},
+					},
+					Output: buildapi.BuildOutput{
+						To: &kapi.ObjectReference{
+							Kind: "DockerImage",
+							Name: "repository/data",
+						},
+					},
+				},
+			},
+		},
+	}
+	for _, buildConfig := range buildConfigs {
+		if result := ValidateBuildConfig(&buildConfig); len(result) > 0 {
+			t.Errorf("Unexpected validation error returned %v", result)
+		}
+	}
+
+	badBuildConfig := buildapi.BuildConfig{
+		ObjectMeta: kapi.ObjectMeta{Name: "config-id", Namespace: "namespace"},
+		Spec: buildapi.BuildConfigSpec{
+			BuildSpec: buildapi.BuildSpec{
+				Source: buildapi.BuildSource{},
+				Strategy: buildapi.BuildStrategy{
+					DockerStrategy: &buildapi.DockerBuildStrategy{},
+				},
+				Output: buildapi.BuildOutput{
+					To: &kapi.ObjectReference{
+						Kind: "DockerImage",
+						Name: "repository/data",
+					},
+				},
+			},
+		},
+	}
+	if result := ValidateBuildConfig(&badBuildConfig); len(result) == 0 {
+		t.Error("An error should have occurred with a DockerStrategy / no source combo")
```

Comment:
this test should be more specific about the validation error it's expecting, like we do elsewhere when testing validation errors.